cloudinit/distros/__init__.py

```diff
@@ -720,6 +725,104 @@ def create_group(self, name, members=None):
                 LOG.info("Added user '%s' to group '%s'", member, name)
 
 
+def _apply_hostname_transformations_to_url(
+    url: str, transformations: "List[Callable[[str], Optional[str]]]"
+) -> "Optional[str]":
+    """
+    Apply transformations to a URL's hostname, return transformed URL.
+
+    This is a separate function because unwrapping and rewrapping only the
+    hostname portion of a URL is complex.
+
+    :param url:
+        The URL to operate on.
+    :param transformations:
+        A list of ``(str) -> Optional[str]`` functions, which will be applied
+        in order to the hostname portion of the URL.  If any function
+        (regardless of ordering) returns None, ``url`` will be returned without
+        any modification.
+
+    :return:
+        A string whose value is ``url`` with the hostname ``transformations``
+        applied, or ``None`` if ``url`` is unparseable.
```

Comment:
If we return `url` it *may* not have transformations applied.  This leaves the caller having to compare the return URL to what it supplied to determine if the transformations were applied or not.  

Would it be better to return None for non-transformable URLs as well?

Reply:
I wrote this before `transformations` _could_ not be applied (by one of them returning `None`), so it could certainly use updating.

That said, I think there are still two cases here (a) the given `transformations` have been _executed_, or (b) the URL is unparseable. I think the issue with the current wording is "applied", as `transformations` can be successfully executed without any transformations being _applied_.

The transformations are not guaranteed to render any changes to the input (and, in fact, in most cases I expect they won't change the input), so if the caller wants to know if the URL has been transformed, they still need to perform a comparison.

Reply:
OK